url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/render/Rasterizer2D.java

```diff
@@ -0,0 +1,458 @@
+    {
+        double dx, dy;
+        double dxdy;
+        double dydx;
+        int x, y;
+        double xx, yy;
+
+        dx = (double)(x1-x0);
+        dy = (double)(y1-y0);
+
+        if ( Math.abs(dx) > VSDK.EPSILON && Math.abs(dy/dx) <= 1 && x1 > x0 ) {
+            // Pendiente entre -1 y 1
+            dydx = dy/dx;
+            for ( x = x0, yy = (double)y0; x <= x1; x++ ) {
+                y = (int)yy;
+                if ( x >= 0 && x < img.getXSize() &&
+                     y >= 0 && y < img.getYSize() ) {
+                    img.putPixelRgb(x, y, p);
+                }
+                yy += dydx;
+            }
+          }
+          else if ( Math.abs(dx) > VSDK.EPSILON && Math.abs(dy/dx) <= 1 && x1 < x0 ) {
+            // Pendiente entre -1 y 1
+            dydx = dy/dx;
+            for ( x = x1, yy = (double)y1; x <= x0; x++ ) {
+                y = (int)yy;
+                if ( x >= 0 && x < img.getXSize() &&
+                     y >= 0 && y < img.getYSize() ) {
+                    img.putPixelRgb(x, y, p);
+                }
+                yy += dydx;
+            }
+          }
+          else if ( Math.abs(dy) > VSDK.EPSILON && y1 > y0 ) {
+            // Pendiente mayor a 1 o menor a -1
+            dxdy = dx/dy;
+            for ( y = y0, xx = (double)x0; y <= y1; y++ ) {
+                x = (int)xx;
+                if ( x >= 0 && x < img.getXSize() &&
+                     y >= 0 && y < img.getYSize() ) {
+                    img.putPixelRgb(x, y, p);
+                }
+                xx += dxdy;
+            }
+          }
+          else if ( Math.abs(dy) > VSDK.EPSILON && y1 < y0 ) {
+            // Pendiente mayor a 1 o menor a -1
+            dxdy = dx/dy;
+            for ( y = y1, xx = (double)x1; y <= y0; y++ ) {
+                x = (int)xx;
+                if ( x >= 0 && x < img.getXSize() &&
+                     y >= 0 && y < img.getYSize() ) {
+                    img.putPixelRgb(x, y, p);
+                }
+                xx += dxdy;
+            }
+          }
+        ;
+    }
+
+    /**
+    Draws a lines outline from a given polygon.
+    */
+    public static void drawPolygon(Image img, Polygon2D p, RGBPixel color)
+    {
+        _Polygon2DContour l;
+        Vertex2D va = null;
+        Vertex2D vb = null;
+        int i;
+        int j;
+
+        for ( i = 0; i < p.loops.size(); i++ ) {
+            for ( j = 0; j < p.loops.get(i).vertices.size() - 1; j++ ) {
+                va = p.loops.get(i).vertices.get(j);
+                vb = p.loops.get(i).vertices.get(j + 1);
+                drawLine(img, (int)va.x, (int)va.y,
+                              (int)vb.x, (int)vb.y, color);
+            }
+            va = vb;
+            vb = p.loops.get(i).vertices.get(0);
+            drawLine(img, (int)va.x, (int)va.y,
+                          (int)vb.x, (int)vb.y, color);
+        }
+    }
+
+    private static void fillPolygonProcessLine(Vertex2D va, Vertex2D vb,
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 20. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 132-154, lines 156-178 into separate methods.